### ISSUE TITLE
Checkout: Update displayed currency and payment methods when the user selects a new country

### ISSUE THREAD

**Issue by DavidRothstein** [Automattic/wp-calypso#22033]
I am working on a project to ensure that users always see the correct currency and payment methods for the country they are purchasing from.

I will use this issue as a central place to track the work.

#### Context

Currently, on a payment page such as this (```/checkout/<domain>/premium```)...

![checkout-page](https://user-images.githubusercontent.com/235183/35661004-5e5214ae-06dd-11e8-829f-48410f84af6a.png)

...the user's currency (US dollars, in the above example) is hardcoded based on the user's auto-detected location. However, if the user is actually billing the purchase to an address in a different country, there is no way for them to see the currency associated with that country instead. For example, if you live in Germany but are traveling in the United States when you make your purchase, you aren't able to see the purchase price in Euros, even after selecting Germany from the country dropdown.

Similarly, the payment options are hardcoded based on the auto-detected location as well; the above screenshot shows credit/debit and PayPal, but some countries get other options here, such as Giropay for Germany.  A user who is in the United States but actually wants to bill to an address in Germany should presumably be able to access the Giropay option, but currently can't.

#### Steps to reproduce
1. Start with a free site, then click "Plans" and choose one of the options and click "Upgrade".
2. No matter what you do on that page (choose a different country in the "Select Country" dropdown, etc.) you'll always see the same currency and available payment methods that you started out with.

The same thing happens for any other type of payment you can make on the site (renewals, add-ons, etc).

#### Plan to fix this

Here is the rough plan I have developed to fix this:

1. **Synchronize the country dropdown between payment methods (and perhaps give it a smart default too).** That is, if you select "Germany" on the credit/debit tab, it should automatically switch to Germany on the PayPal tab too. (And perhaps also, default it to Germany in the first place if that is the user's auto-detected location.)  This would be a small usability enhancement in its own right, but it also contributes to the main goal here, because if the country dropdown is going to determine the appearance of other things on the page, then the UI should be consistent about which country is used to do that.
**Status: Mostly done in https://github.com/Automattic/wp-calypso/pull/22239, with the "smart default" based on the user's auto-detected location done in https://github.com/Automattic/wp-calypso/pull/23810.**
2. **When you change the country dropdown during checkout, update everything on the page to reflect the new currency and payment options.**  On the client side, this will be achieved by building off the above work to ensure that the shopping cart is aware of the current payment country and triggers a cart update when it changes.  This will also likely require small changes to the WordPress.com REST API to allow a specific country/currency to be passed in so that information about the shopping cart/plans/etc. comes back with the correct currency.
**Status: Done in https://github.com/Automattic/wp-calypso/pull/23810, but it still needs some polish and is also waiting on some WordPress.com REST API changes to get in.**
3. **Figure out what else needs to change during checkout and the implications of doing so.**  There are other places that display a product's currency or have a country dropdown also. The stated goal of this project was to change a couple specific ones, but I have to figure out exactly where to draw the line. For example, the shopping cart in the above screenshot clearly needs to have its currency auto-updated, but the same shopping cart is visible at ```/plans/<domain>``` also (if the user abandons the checkout process and then navigates back to the plans page later):
![plans-page](https://user-images.githubusercontent.com/235183/35661005-63cb7d12-06dd-11e8-88a4-f101d63f2890.png)
Assuming the shopping cart should get updated here as well (presumably it will automatically), then the Personal/Premium prices listed on this page would ideally be updated to match... but if we start updating everything around the UI based on the user's selection, it is a little odd to be doing that based on a particular dropdown within the payment workflow (a centralized place to select the country beforehand would make more sense instead).  So, this requires some more thought to carefully define the scope of this issue.
**[Conclusion for now: As discussed in the issue comments, for now we'll only update currencies in the shopping cart and anything derived directly from it, like the "Pay $96" button.  We won't try to address the edge case discussed above regarding plan prices, since it impacts parts of the site that are independent of the specific purchase that is currently being made.]**
4. **Synchronize the country dropdown in other locations where the payment country is selected outside of the final checkout payment page (such as the add/edit credit card forms found via ```/me/purchases```, and the domain contact info form during checkout).** The add/edit credit card forms are already automatically handled via the first bullet point above. For the domain contact form, this will require a separate pull request, since (a) it can already have its own defaults for all fields including the country (based on the last time the user registered a domain), and (b) it is not clear that we should use the country selected on this form to override the default country from any previously-entered payment methods (although we definitely should use it to set an initial default payment country if there wasn't one before).
**Status: Done in https://github.com/Automattic/wp-calypso/pull/23810, but may still need some work for part (b).**

**Comment by rralian** [comment]
Point number 3 is interesting and I'd need to think about it a bit more. I believe currency is an attribute stored at the user level (possibly independent of the forms we're talking about here). But you _could_ have two different payments methods associated with different subscriptions, and those payment methods could come from two different countries. I'm not actually sure about the logic behind the user-level currency attribute and whether/when we'd want to update that and whether that should dictate the displayed currencies in other areas or whether that should be determined instead by geoip/locale + user-selection. @yoavf probably has some smart things to say here. But generally we'd be pretty happy just making the payment form dynamically honor the user's country selection.

**Comment by DavidRothstein** [comment]
@rralian Thank you for the feedback.

> I believe currency is an attribute stored at the user level (possibly independent of the forms we're talking about here).

The closest thing returned by the [current user API](https://developer.wordpress.com/docs/api/1.1/get/me/) seems to be ```user_ip_country_code``` (and Calypso doesn't look like it even uses that property; it gets the country via a separate API call).  But I suspect you are right that behind the scenes, this is what is used to determine the currency that the other API methods (cart, plans, etc.) send back.

> But you could have two different payments methods associated with different subscriptions, and those payment methods could come from two different countries. I'm not actually sure about the logic behind the user-level currency attribute and whether/when we'd want to update that... But generally we'd be pretty happy just making the payment form dynamically honor the user's country selection.

Interesting point.  OK, I think that argues for drawing the initial line here at just updating the currency in the shopping cart (and anything derived directly from the cart, like the "Pay $96" button text), but nowhere else, at least for now.  I think if I implement this correctly, then changing the currency in one of those places will automatically change it in the others; I will have a better idea when I come back later today and look at this more.

This does mean that in my second screenshot above, you could wind up with a scenario where the plans display one currency and the shopping cart displays a different currency, but this is kind of an edge case.

**Comment by DavidRothstein** [comment]
Initial pull request associated with part 1 of this issue is at https://github.com/Automattic/wp-calypso/pull/22239

**Comment by DavidRothstein** [comment]
Final pull request for this issue is now at https://github.com/Automattic/wp-calypso/pull/23810 (although it still needs some polish and is also waiting on some WordPress.com REST API changes to get in).

**Comment by DavidRothstein** [comment]
This was my Automattic trial project, and It's OK to close it.

We wound up implementing part of it in Calypso, and making a number of changes and improvements to the code underlying the REST API to help support the rest of it.  However, we decided a while ago not to go ahead and expose this feature to end users for now, I think because people were concerned about extra complexity to support what is somewhat of an edge case.
